### PR TITLE
add update_position in YASPool

### PR DESCRIPTION
**Pull request by dubzn**
Closes: #87

### CODE REVIEW

File: src/tests/test_contracts/test_yas_pool.cairo [hunk 1]
```diff
@@ -198,4 +206,205 @@ mod YASPoolTests {
...
+
+    fn mock_ticks(ref tick_state: Tick::ContractState, from: u32, to: u32, info: Tick::Info) {
+        assert(from <= to, 'init_mock_ticks - from > to');
+
+        let mut i = from;
+        loop {
+            if i == to + 1 {
+                break;
+            }
+            Tick::InternalImpl::set_tick(ref tick_state, IntegerTrait::<i32>::new(i, false), info);
```
Comment (rcatalan98) [inline src/tests/test_contracts/test_yas_pool.cairo:406]: Another option would be to send a `Snap` and then do the iteration in `set_tick`

File: src/contracts/yas_pool.cairo [hunk 2]
```diff
@@ -433,10 +434,149 @@ mod YASPool {
...
+            let _fee_growth_global_1_X128 = self.fee_growth_global_1_X128.read();
+
+            let _max_liquidity_per_tick = self.max_liquidity_per_tick.read();
+
+            // if we need to update the ticks, do it
+            let mut flipped_lower = false;
+            let mut flipped_upper = false;
+
+            if liquidity_delta.is_non_zero() {
+                // TODO: modify block_time from Tick::update()
```
[the remark this answers is not in the document]
Comment (dubzn) [reply src/contracts/yas_pool.cairo:460]: Already did it in this PR, but forgot to remove the TODO haha :heavy_check_mark:

File: src/contracts/yas_pool.cairo [hunk 3]
```diff
@@ -433,10 +434,149 @@ mod YASPool {
...
+        /// @param tick_upper the upper tick of the position's tick range
+        /// @param tick the current tick, passed to avoid sloads
+        fn update_position(
+            self: @ContractState, position_key: PositionKey, liquidity_delta: i128, tick: i32
+        ) -> Info {
+            let mut tick_bitmap_state = TickBitmap::unsafe_new_contract_state();
+            let mut tick_state = Tick::unsafe_new_contract_state();
+            let mut position_state = Position::unsafe_new_contract_state();
+
+            // SLOAD for gas optimization
```
[the remark this answers is not in the document]
Comment (dubzn) [reply src/contracts/yas_pool.cairo:449]: done :heavy_check_mark:

File: src/contracts/yas_pool.cairo [hunk 4]
```diff
@@ -433,10 +434,149 @@ mod YASPool {
...
+        }
+
+        fn get_tick_state(self: ContractState) -> Tick::ContractState {
+            Tick::unsafe_new_contract_state()
+        }
+
+        fn get_position_state(self: ContractState) -> Position::ContractState {
+            Position::unsafe_new_contract_state()
+        }
+
```
[the remark this answers is not in the document]
Comment (dubzn) [reply src/contracts/yas_pool.cairo:551]: done :heavy_check_mark:

File: src/contracts/yas_pool.cairo [hunk 5]
```diff
@@ -433,10 +434,149 @@ mod YASPool {
...
+        ) -> Info {
+            let mut tick_bitmap_state = TickBitmap::unsafe_new_contract_state();
+            let mut tick_state = Tick::unsafe_new_contract_state();
+            let mut position_state = Position::unsafe_new_contract_state();
+
+            // SLOAD for gas optimization
+            let _fee_growth_global_0_X128 = self.fee_growth_global_0_X128.read();
+            let _fee_growth_global_1_X128 = self.fee_growth_global_1_X128.read();
+
+            let _max_liquidity_per_tick = self.max_liquidity_per_tick.read();
```
Comment (dpinones) [inline src/contracts/yas_pool.cairo:453]:
```suggestion
            let fee_growth_global_0_X128 = self.fee_growth_global_0_X128.read();
            let fee_growth_global_1_X128 = self.fee_growth_global_1_X128.read();

            let max_liquidity_per_tick = self.max_liquidity_per_tick.read();
```

Comment (dubzn) [reply src/contracts/yas_pool.cairo:453]: done :heavy_check_mark:

File: src/tests/test_contracts/test_yas_pool.cairo (hunk 1 above)
Comment (dubzn) [reply src/tests/test_contracts/test_yas_pool.cairo]: Just added a method in the `InternalImpl` of `Tick`, which is `set_ticks()`, and it takes an array of ticks and infos. Check it out please

File: src/libraries/tick.cairo
[the remark this answers is not in the document]
Comment (dubzn) [reply src/libraries/tick.cairo:1]: I added a comment in `update()` method in Tick contract.

File: src/libraries/tick.cairo [hunk 6]
```diff
@@ -282,6 +276,18 @@ mod Tick {
...
+            assert(ticks.len() == infos.len(), 'ticks & infos must have eq len');
+            let mut i = 0;
+            loop {
+                if i > ticks.len() - 1 {
+                    break;
+                }
+                InternalImpl::set_tick(ref self, *(ticks.at(i)), *(infos.at(i)));
+                i += 1;
+            };
+        }
```
Comment (rcatalan98) [inline src/libraries/tick.cairo:289]:
```suggestion
        fn set_ticks(ref self: ContractState, ticks: Array<i32>, ticks_info: Array<Info>) {
            assert(ticks.len() == ticks_info.len(), 'ticks & infos must have eq len');
            let mut i = 0;
            loop {
                if i > ticks.len() - 1 {
                    break;
                }
                InternalImpl::set_tick(ref self, *(ticks.at(i)), *(ticks_info.at(i)));
                i += 1;
            };
        }
```

**Review by rcatalan98 — CHANGES_REQUESTED** [review]
Just a really small refactor that would add more clarity. But great work!

**Review by rcatalan98 — APPROVED** [review]
LGTM!

**Review by dpinones — APPROVED** [review]

**Pull request merged**